app/src/main/kotlin/compiler/parser/analysis/GrammarAnalysis.kt

```diff
@@ -12,6 +18,89 @@ class GrammarAnalysis<S : Comparable<S>> {
     }
 
     fun computeFollow(grammar: AutomatonGrammar<S>, nullable: Set<S>, first: Map<S, Set<S>>): Map<S, Set<S>> {
-        return TODO()
+
+        val result: MutableMap<S, MutableSet<S>> = HashMap()
+
+        var wasResultUpdated = true
+        fun update(setToUpdate: MutableSet<S>, secondSet: Set<S>) {
+            val sizeBefore = setToUpdate.size
+            setToUpdate.addAll(secondSet)
+            if (sizeBefore < setToUpdate.size) wasResultUpdated = true
+        }
+
+        fun initializeResult() {
+            for (lhs in grammar.productions.keys) result[lhs] = HashSet()
+            val statesFirstMap: MutableMap<S, MutableMap<DfaState<S, Production<S>>, MutableSet<S>>> = HashMap()
+            while (wasResultUpdated) {
+                wasResultUpdated = false
+                for ((lhs, dfa) in grammar.productions) {
+                    val statesFirst = statesFirstMap.getOrPut(lhs) { HashMap() }
+                    val acceptingStates = dfa.getAcceptingStates()
+                    val visited: MutableSet<DfaState<S, Production<S>>> = acceptingStates.toHashSet()
+                    val queue: Queue<DfaState<S, Production<S>>> = LinkedList()
+                    queue.addAll(acceptingStates)
+
+                    while (!queue.isEmpty()) {
```

Comment:
Similar bfs logic for accepting states backwards is used here twice and also for nullable, maybe we can create abstract bfs with action lambda as argument and use it to avoid repetition.

Reply:
Maybe this one will be easier to do after merging